src/trace.zig

```diff
@@ -112,304 +112,324 @@ pub const EventMetricAggregate = @import("trace/event.zig").EventMetricAggregate
-            "\"ph\":\"{[event]c}\"," ++
-            "\"ts\":{[timestamp]}," ++
-            "\"cat\":\"{[category]s}\"," ++
-            "\"name\":\"{[category]s} {[event_tracing]} {[event_timing]}\"," ++
-            "\"args\":{[args]s}" ++
-            "}},\n", .{
-            .process_id = tracer.replica_index,
-            .thread_id = event_tracing.stack(),
-            .category = @tagName(event),
-            .event = 'B',
-            .timestamp = time_elapsed.microseconds(),
-            .event_tracing = event_tracing,
-            .event_timing = event_timing,
-            .args = std.json.Formatter(Event){ .value = event, .options = .{} },
-        }) catch {
-            log.err("{}: {s}({}): event too large: {}", .{
-                tracer.replica_index,
-                @tagName(event),
-                event_tracing,
-                event_timing,
-            });
-            return;
-        };
+            return .{
+                .time = time,
+                .replica_index = replica_index,
+                .options = options,
+                .buffer = buffer,
+                .statsd = statsd,
 
-        writer.writeAll(buffer_stream.getWritten()) catch |err| {
-            std.debug.panic("Tracer.start: {}\n", .{err});
-        };
-    }
-
-    pub fn stop(tracer: *Tracer, event: Event) void {
-        const us_log_threshold_ns = 5 * std.time.ns_per_ms;
-
-        const event_tracing = event.as(EventTracing);
-        const event_timing = event.as(EventTiming);
-        const stack = event_tracing.stack();
-
-        const event_start = tracer.events_started[stack].?;
-        const event_end = stdx.Instant.now();
-        const event_duration = event_end.duration_since(event_start);
-
-        assert(tracer.events_started[stack] != null);
-        tracer.events_started[stack] = null;
-
-        // Double leading space to align with 'start: '.
-        log.debug("{}: {s}({}): stop:  {} (duration={}{s})", .{
-            tracer.replica_index,
-            @tagName(event),
-            event_tracing,
-            event_timing,
-            if (event_duration.nanoseconds < us_log_threshold_ns)
-                event_duration.microseconds()
-            else
-                event_duration.milliseconds(),
-            if (event_duration.nanoseconds < us_log_threshold_ns) "us" else "ms",
-        });
-
-        tracer.timing(event_timing, event_duration.microseconds());
-
-        tracer.write_stop(stack, event_duration);
-    }
-
-    pub fn cancel(tracer: *Tracer, event_tag: Event.Tag) void {
-        const stack_base = EventTracing.stack_bases.get(event_tag);
-        const cardinality = EventTracing.stack_limits.get(event_tag);
-        for (stack_base..stack_base + cardinality) |stack| {
-            if (tracer.events_started[stack]) |event_start| {
-                log.debug("{}: {s}: cancel", .{ tracer.replica_index, @tagName(event_tag) });
-
-                const event_end = stdx.Instant.now();
```

Comment:
It is a pet-peeve of mine where `.now()` is called several times in the same function. I like to cache such reads, not for performance, but to make sure that things that logically happened at the same time get the same physical timestamp.